packages/client/src/runtime/core/protocol/graphql.ts

```diff
@@ -0,0 +1,138 @@
+import { EngineQuery } from '@prisma/engine-core'
+
+import { DMMFHelper } from '../../dmmf'
+import { ErrorFormat } from '../../getPrismaClient'
+import { Args, Document, makeDocument, unpack } from '../../query'
+import { Action } from '../types/JsApi'
+import { CreateMessageOptions, ProtocolEncoder, ProtocolMessage } from './common'
+
+const actionOperationMap: Record<Action, 'query' | 'mutation'> = {
+  findUnique: 'query',
+  findUniqueOrThrow: 'query',
+  findFirst: 'query',
+  findFirstOrThrow: 'query',
+  findMany: 'query',
+  count: 'query',
+  create: 'mutation',
+  createMany: 'mutation',
+  update: 'mutation',
+  updateMany: 'mutation',
+  upsert: 'mutation',
+  delete: 'mutation',
+  deleteMany: 'mutation',
+  executeRaw: 'mutation',
+  queryRaw: 'mutation',
+  aggregate: 'query',
+  groupBy: 'query',
+  runCommandRaw: 'mutation',
+  findRaw: 'query',
+  aggregateRaw: 'query',
+}
+
+export class GraphQLProtocolEncoder implements ProtocolEncoder {
+  constructor(private dmmf: DMMFHelper, private errorFormat: ErrorFormat) {}
+
+  createMessage({ action, model, args, extensions, clientMethod, callsite }: CreateMessageOptions): ProtocolMessage {
+    let rootField: string | undefined
+    const operation = actionOperationMap[action]
+
+    if (action === 'executeRaw' || action === 'queryRaw' || action === 'runCommandRaw') {
+      rootField = action
+    }
+
+    let mapping
+    if (model !== undefined) {
+      mapping = this.dmmf?.mappingsMap[model]
+      if (mapping === undefined) {
+        throw new Error(`Could not find mapping for model ${model}`)
+      }
+
+      rootField = mapping[action === 'count' ? 'aggregate' : action]
+    }
+
+    if (operation !== 'query' && operation !== 'mutation') {
+      throw new Error(`Invalid operation ${operation} for action ${action}`)
+    }
+
+    const field = this.dmmf?.rootFieldMap[rootField!]
+
+    if (field === undefined) {
+      throw new Error(
+        `Could not find rootField ${rootField} for action ${action} for model ${model} on rootType ${operation}`,
+      )
+    }
+
+    // TODO: figure out if can be done in PrismaClient iteself
```

Comment:
Is the model name enough to make this work for both implementations?

Reply:
I think this needs to be resolved before you merge?

Reply:
Oh. actually, I already did this even before I opened PR, just forgot to remove this comment.